aten/src/ATen/native/cuda/BatchLinearAlgebra.cu

```diff
@@ -2566,95 +2566,149 @@ std::tuple<Tensor, Tensor, Tensor> _svd_helper_cuda(const Tensor& self, bool som
-        break;
-      }
-    }
+  auto b_stride = matrixStride(b);
+  auto lu_stride = matrixStride(lu);
+  auto pivots_stride = pivots.size(-1);
+  magma_int_t batch_size = magma_int_cast(batchCount(b), "batchCount");
 
-    // Compute whatever is left = batch_size - floor(batch_size / batch_limit) * batch_limit
-    // which concisely is equal to batch_size % batch_limit
-    if (batch_size % batch_limit != 0 && info_tmp == 0) {
-      magmaLuSolveBatched<scalar_t>(
-          n, nrhs, &lu_array[mini_idx], n, &pivots_array[mini_idx], &b_array[mini_idx], n,
-          info_tmp, batch_size % batch_limit, magma_queue);
-    }
+  magma_int_t** pivots_array;
+  scalar_t** lu_array;
+  scalar_t** b_array;
 
-    info = info_tmp;
+  ALLOCATE_ARRAY(pivots_array, magma_int_t*, batch_size);
+  ALLOCATE_ARRAY(lu_array, scalar_t*, batch_size);
+  ALLOCATE_ARRAY(b_array, scalar_t*, batch_size);
+
+  for (int64_t i = 0; i < batch_size; i++) {
+    pivots_array[i] = &pivots_data[i * pivots_stride];
+    b_array[i] = &b_data[i * b_stride];
+    lu_array[i] = &lu_data[i * lu_stride];
+  }
+
+  MAGMAQueue magma_queue(b.get_device());
+
+  // Compute the result in batches of 65535
+  // that is the maximum allowed number for batch_size in MAGMA
+  constexpr int64_t batch_limit = 65535;
+
+  for (int64_t mini_idx = 0; mini_idx < batch_size; mini_idx += batch_limit) {
+    int64_t nbatches = std::min(batch_limit, batch_size - mini_idx);
+    scalar_t** lu_array_cur = &lu_array[mini_idx];
+    scalar_t** b_array_cur = &b_array[mini_idx];
+    magma_int_t** pivots_array_cur = &pivots_array[mini_idx];
+
+    int info;
+    magmaLuSolveBatched<scalar_t>(
+        n, nrhs, lu_array_cur, leading_dimension,
+        pivots_array_cur, b_array_cur, leading_dimension,
+        info, nbatches, magma_queue);
+
+    // info from magmaLuSolveBatched only reports if the i-th parameter is wrong
+    // so we don't need to check it all the time
+    TORCH_INTERNAL_ASSERT_DEBUG_ONLY(info == 0);
   }
 #endif
 }
 
-Tensor _lu_solve_helper_cuda(const Tensor& self, const Tensor& LU_data, const Tensor& LU_pivots) {
-  int64_t info = 0;
-  auto self_working_copy = cloneBatchedColumnMajor(self);
-  auto LU_data_working_copy = cloneBatchedColumnMajor(LU_data);
-  auto LU_pivots_working_copy = LU_pivots.is_contiguous() ? LU_pivots : LU_pivots.contiguous();
-
-  if (self.numel() == 0 || LU_data.numel() == 0) {
-    return at::zeros_like(self, LEGACY_CONTIGUOUS_MEMORY_FORMAT);
+static void lu_solve_magma(const Tensor& b, const Tensor& lu, const Tensor& pivots) {
+  // TODO: compare performance and use the best performing option based on lu's sizes
+  if (b.dim() == 2) {
+    AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES(b.scalar_type(), "lu_solve_magma", [&]{
+      apply_lu_solve_looped_magma<scalar_t>(b, lu, pivots);
+    });
+  } else {
+    AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES(b.scalar_type(), "lu_solve_magma", [&]{
+      apply_lu_solve_batched_magma<scalar_t>(b, lu, pivots);
+    });
```

Comment:
The internal comments on batched in looped versions mention that the batched version is faster than the looped for small input, i.e.
```
/*
  Solves the matrix equation A X = B
  X and B are n-by-nrhs matrices, A is represented using the LU factorization.
  This is an in-place routine, content of `b` is overwritten.
  This is a specialized batched variant, it is expected to be faster than the "looped" version only for small inputs.
```
In this code, however, the batched version is always chosen for a batched input of arbitrary size.

Reply:
I kept the original conditions for choosing the single input variant over the batched one. A better condition that is based on the size of the input will be added separately in the future.
cc: @v0dro